### PR TITLE
Add default model for the nvidia classes in local NIM mode

### PR DESCRIPTION
**Pull request by raspawar**
# Description

Provide default model using the first available model if none is provided for local NIM, e.g.
Settings.llm = NVIDIA(base_url="http://localhost:1234/v1")
query_engine = index.as_query_engine(similarity_top_k=20)
response = query_engine.query(
"How many new housing units were built in San Francisco in 2021?"
)
llm._client.model => default model => first available model in local NIM
Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [X] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [X] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Added new unit/integration tests
- [X] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I ran `make format; make lint` to appease the lint gods

@sumitkbh

### CODE REVIEW

**Comment by raspawar** [comment]
@mattf please have a look

**Comment by logan-markewich** [comment]
Looks like this maybe broke tests?

**Review by mattf — COMMENTED** [review]
@raspawar Logan is right. i just checked main vs this PR. main passes and PR fails -

```
FAILED tests/test_api_key.py::test_missing_api_key_error - AssertionError: assert '401' in '404 page not found'
FAILED tests/test_api_key.py::test_bogus_api_key_error - AssertionError: assert '401' in '404 page not found'
FAILED tests/test_embeddings_nvidia.py::test_nvidia_embedding_callback - openai.NotFoundError: 404 page not found
FAILED tests/test_embeddings_nvidia.py::test_nvidia_embedding_throws_with_invalid_key - openai.NotFoundError: 404 page not found
```

the 404 suggests the model name is being mangled, but in my testing it may be an issue w/ using the openai client.

```
>>> import os
>>> from openai import OpenAI
>>> client = OpenAI(api_key=os.environ["NVIDIA_API_KEY"], base_url="https://integrate.api.nvidia.com/v1")
>>> x = client.embeddings.create(input=['hello'], model="NV-Embed-QA", extra_body={"input_type": "query"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/resources/embeddings.py", line 114, in create
    return self._post(
  File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 1266, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
  File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 942, in request
    return self._request(
  File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 1046, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.NotFoundError: 404 page not found
```

**Comment by raspawar** [comment]
> @raspawar Logan is right. i just checked main vs this PR. main passes and PR fails -
> 
> ```
> FAILED tests/test_api_key.py::test_missing_api_key_error - AssertionError: assert '401' in '404 page not found'
> FAILED tests/test_api_key.py::test_bogus_api_key_error - AssertionError: assert '401' in '404 page not found'
> FAILED tests/test_embeddings_nvidia.py::test_nvidia_embedding_callback - openai.NotFoundError: 404 page not found
> FAILED tests/test_embeddings_nvidia.py::test_nvidia_embedding_throws_with_invalid_key - openai.NotFoundError: 404 page not found
> ```
> 
> the 404 suggests the model name is being mangled, but in my testing it may be an issue w/ using the openai client.
> 
> ```
> >>> import os
> >>> from openai import OpenAI
> >>> client = OpenAI(api_key=os.environ["NVIDIA_API_KEY"], base_url="https://integrate.api.nvidia.com/v1")
> >>> x = client.embeddings.create(input=['hello'], model="NV-Embed-QA", extra_body={"input_type": "query"})
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/resources/embeddings.py", line 114, in create
>     return self._post(
>   File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 1266, in post
>     return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
>   File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 942, in request
>     return self._request(
>   File "/home/matt/.cache/pypoetry/virtualenvs/llama-index-embeddings-nvidia-LEYbhevy-py3.10/lib/python3.10/site-packages/openai/_base_client.py", line 1046, in _request
>     raise self._make_status_error_from_response(err.response) from None
> openai.NotFoundError: 404 page not found
> ```

I will investigate this

**Comment by raspawar** [comment]
@logan-markewich I have mocked the failing test case API calls, can you ptal?

**Review by mattf — COMMENTED** [review]
please add a test for the embedding interface, then ok to merge

**Comment by mattf** [comment]
:+1:

**Review by logan-markewich — APPROVED** [review]

File: llama-index-integrations/postprocessor/llama-index-postprocessor-nvidia-rerank/poetry.lock
Comment (logan-markewich) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-nvidia-rerank/poetry.lock:1]: please don't commit the lock files, they just take up space

**Comment by logan-markewich** [comment]
Ok @raspawar fixed one issue with tests, now it's up to you :)

**Pull request merged**